agent/lib_php_amqplib.c

```diff
@@ -0,0 +1,401 @@
+  nr_php_amqplib_get_host_and_port(amqp_connection, &message_params);
+
+  /* Compatibility with PHP 7.x */
+  NR_PHP_WRAPPER_CALL;
+
+  if (NULL == auto_segment) {
+    /*
+     * Must be checked after PHP_WRAPPER_CALL to ensure txn didn't end during
+     * the call.
+     */
+    goto end;
+  }
+  /*
+   *The retval should be an AMQPMessage. nr_php_is_zval_* ops do NULL checks
+   * as well.
+   */
+  if (nr_php_is_zval_valid_object(*retval_ptr)) {
+    /*
+     *  Get the exchange and routing key from the AMQPMessage
+     */
+    amqp_exchange = nr_php_get_zval_object_property(*retval_ptr, "exchange");
+    if (nr_php_is_zval_non_empty_string(amqp_exchange)) {
+      /* Used with consumer only; his is exchange name.  Exchange name is
+       * Default in case of empty string. */
+      message_params.messaging_destination_publish_name
+          = Z_STRVAL_P(amqp_exchange);
+    } else {
+      message_params.messaging_destination_publish_name = "Default";
+    }
+
+    amqp_routing_key
+        = nr_php_get_zval_object_property(*retval_ptr, "routingKey");
+    if (nr_php_is_zval_non_empty_string(amqp_routing_key)) {
+      message_params.messaging_destination_routing_key
+          = Z_STRVAL_P(amqp_routing_key);
+    }
+  }
+
+  /* Now create and end the instrumented segment as a message segment. */
+  /*
+   * By this point, it's been determined that this call will be instrumented so
+   * only create the message segment now, grab the parent segment start time,
+   * add our message segment attributes/metrics then close the newly created
+   * message segment.
+   */
+  message_segment = nr_segment_start(NRPRG(txn), NULL, NULL);
+
+  if (NULL == message_segment) {
+    goto end;
+  }
+
+  /* re-use start time from auto_segment started in func_begin */
+  message_segment->start_time = auto_segment->start_time;
+
+  nr_segment_message_end(&message_segment, &message_params);
+
+end:
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO
+    /* PHP 8+ */
+    /* gnu compiler needed closure. */
+    ;
+#else
+  /*
+   * Because we had to strdup values to persist them beyond NR_PHP_WRAPPER_CALL,
+   * now we destroy them. There isn't a separate function to destroy all since
+   * some of the params are string literals and we don't want to strdup
+   * everything if we don't have to. RabbitMQ basic_get PHP 7.x will only strdup
+   * server_address and destination_name.
+   */
+  nr_free(message_params.server_address);
+  nr_free(message_params.destination_name);
+#endif
+}
+NR_PHP_WRAPPER_END
+
+void nr_php_amqplib_enable() {
+  /*
+   * Set the UNKNOWN package first, so it doesn't overwrite what we find with
+   * nr_php_amqplib_handle_version.
+   */
+  if (NRINI(vulnerability_management_package_detection_enabled)) {
+    nr_txn_add_php_package(NRPRG(txn), PHP_PACKAGE_NAME,
+                           PHP_PACKAGE_VERSION_UNKNOWN);
+  }
+
+  /* Extract the version for aws-sdk 3+ */
+  nr_php_amqplib_handle_version();
+  nr_php_amqplib_ensure_class();
```

Comment:
I am concerned on how this all works - if the file used to trigger this enable function is, for example, in a sequence of PHP calls that was in response to autoloading the class we need, and inside this sequence we start another sequence to load this class - is composer setup to handle this?

Is there not a file we can use in the situations this class is not triggered (the AWS context I guess?).

Reply:
No, none of the autoload files can guarantee this class being loaded in time for the AWS setup which is either do to the slight variance in an SSL connection or to the older rabbitmq engine that is running.  None of the files that loaded when running AWS tests worked to allow our instrumentation to be wrapped (different than the non-ssl, non-aws rabbitmq server which actually loaded the file as well as the class).

All this is doing is calling `class_exists` so pretty innocuous.  This is a php function which other functions can call in the middle of their and have easily handled in the middle of other function calls as well.  It's analogous to function calls that use other classes.  The class doesn't get loaded until it's used.  Same thing here.